packages/components/src/slot-fill/context.js

```diff
@@ -60,6 +60,9 @@ class SlotFillProvider extends Component {
 		// will be empty (the new Slot will subsume all fills for this name).
 		if ( previousSlot ) {
 			previousSlot.forceUpdate();
+
+			// See 'should re-render Fill when fillProps are updated' test.
+			this.forceUpdateFills( name );
```

Comment:
The issue is that `Fill` wasn't being re-rendered when `fillProps` changed because [`contextValue`](https://github.com/WordPress/gutenberg/blob/c721a4d1daf3176f6118b9261d2821a9322679a7/packages/components/src/slot-fill/context.js#L37-L45) never changes (and therefore there's no need to re-render consumers), even when there are new `fillProps`. In fact, `fillProps` is not included in the context state. [It's accessed using the `Slot` instance](https://github.com/WordPress/gutenberg/blob/c721a4d1daf3176f6118b9261d2821a9322679a7/packages/components/src/slot-fill/fill.js#L58) that is returned by `context.getSlot()`. Both the getter function and the instance reference never change.

My first attempt was transforming `contextValue` into `this.state`, put `fillProps` there and update it when new `fillProps` arrived, which would make consumers re-render. Not sure why it hasn't been done that way in the first place. I'll be glad to refactor `Slot/Fill` in a separate PR as well.

Reply:
Yes, it seems like its own PR would make it easier to review, test and land. It should be quite straightforward to tackle.